### DISCUSSION TITLE
Close modal after server action using revalidatePath completes

### DISCUSSION THREAD

**Discussion by joepetrillo** [vercel/next.js#50559] [App Router]
So, I control if the modal is open/closed using an `isOpen` boolean state.

According to the NextJS docs, if a server action mutates data and calls `redirect`, `revalidatePath`, or `revalidateTag `, you are supposed to use the `useTransition` hook to start the action (not sure why exactly). This hook provides us with an `isPending` boolean that represents if the action is running and a `startTransition` function to start it.

The form inside of the modal should be disabled and show a spinner indicating when the server action is running. (I am ok with the UI being blocked here during this time).

My confusion is how to close the modal once the action is complete. I have it working as expected by just using it like an async function (see screenshot), but apparently it's not the correct way. I am using react-hook-form and shadcn for my form validation so there is an onSubmit function I have available.

![image](https://github.com/vercel/next.js/assets/32227357/d31eb98f-2598-4326-b3f9-8e70f68fdc2b)

**Comment by pi-kei** [comment]
It's almost similar discussion here #49426 
Currently there's no 100% right way to do that and using action as an async function is the way to go.

**Reply by codinginflow** [reply]
> It's almost similar discussion here #49426 Currently there's no 100% right way to do that and using action as an async function is the way to go.

In the example, they are returning an error in the return value of the server action. If we use it like a normal async function, why not simply wrap it in try/catch?

**Comment by codinginflow** [comment]
I have the same confusion. Everyone is just saying "You should use server actions in this way" without explaining why. For example, why the `useTransition`? And why only for mutations?

I am using server actions the same way you do and it works for me as well.
